File: generator/src/main/resources/base/io/apitomy/umg/base/util/JsonUtil.java
package io.apitomy.umg.base.util;

import java.io.IOException;
import java.util.ArrayList;
import java.util.Collection;

import java.util.LinkedList;
import java.util.List;
import java.util.Map;
import java.util.regex.Pattern;
import java.util.stream.Collectors;

import com.fasterxml.jackson.core.JsonGenerator;
import com.fasterxml.jackson.core.util.MinimalPrettyPrinter;
import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.databind.ObjectMapper;
import com.fasterxml.jackson.databind.node.ArrayNode;
import com.fasterxml.jackson.databind.node.JsonNodeFactory;
import com.fasterxml.jackson.databind.node.ObjectNode;
import com.fasterxml.jackson.databind.node.TextNode;
import com.fasterxml.jackson.databind.util.TokenBuffer;

public class JsonUtil {

    private static final JsonNodeFactory factory = JsonNodeFactory.instance;
    private static final ObjectMapper mapper = new ObjectMapper();

    public static List<String> keys(ObjectNode json) {
        List<String> rval = new ArrayList<>();
        if (json != null) {
            json.fieldNames().forEachRemaining(rval::add);
        }
        return rval;
    }

    public static List<String> matchingKeys(String regex, ObjectNode json) {
        return keys(json).stream().filter(key -> Pattern.matches(regex, key)).collect(Collectors.toList());
    }

    public static JsonNode getProperty(ObjectNode json, String propertyName) {
        if (json.has(propertyName)) {
            return json.get(propertyName);
        }
        return null;
    }

    public static void setProperty(ObjectNode json, String propertyName, JsonNode value) {
        if (value != null) {
            json.set(propertyName, value);
        }
    }


    public static String stringify(JsonNode json) {
        try {
            PrettyPrinter pp = new PrettyPrinter();
            return mapper.writer(pp).writeValueAsString(json);
        } catch (Exception e) {
            throw new RuntimeException(e);
        }
    }

    public static JsonNode parseJSON(String jsonString) {
        try {
            return mapper.readTree(jsonString);
        } catch (IOException e) {
            throw new RuntimeException(e);
        }
    }

    public static JsonNode clone(JsonNode json) {
        try {
            TokenBuffer tb = new TokenBuffer(mapper, false);
            mapper.writeTree(tb, json);
            return mapper.readTree(tb.asParser());
        } catch (IOException e) {
            throw new RuntimeException("Error cloning JSON node.", e);
        }
    }

    public static <T> List<T> collectionToList(Collection<T> collection) {
        if (collection == null) {
            return new ArrayList<>();
        }
        return new ArrayList<>(collection);
    }

    public static ObjectNode objectNode() {
        return factory.objectNode();
    }

    public static ArrayNode arrayNode() {
        return factory.arrayNode();
    }

    public static TextNode textNode(String value) {
        return factory.textNode(value);
    }

    public static JsonNode toJsonNode(Object value) {
        if (value == null) {
            return null;
        }
        if (value instanceof JsonNode) {
            return (JsonNode) value;
        }
        if (value instanceof String) {
            return factory.textNode((String) value);
        }
        if (value instanceof Boolean) {
            return factory.booleanNode((Boolean) value);
        }
        if (value instanceof Integer) {
            return factory.numberNode((Integer) value); // IntNode — preserves 42 (not 42.0)
        }
        if (value instanceof Number) {
            return factory.numberNode(((Number) value).doubleValue()); // DoubleNode
        }
        return null;
    }

    public static ArrayNode toArrayNode(List<?> list) {
        if (list == null) {
            return null;
        }
        ArrayNode array = factory.arrayNode(list.size());
        for (int i = 0; i < list.size(); i++) {
            JsonNode node = toJsonNode(list.get(i));
            if (node != null) {
                array.add(node);
            }
        }
        return array;
    }

    public static ObjectNode toObjectNode(Map<String, ?> map) {
        if (map == null) {
            return null;
        }
        ObjectNode object = factory.objectNode();
        List<String> mapKeys = new ArrayList<>(map.keySet());
        for (int i = 0; i < mapKeys.size(); i++) {
            String key = mapKeys.get(i);
            JsonNode node = toJsonNode(map.get(key));
            if (node != null) {
                object.set(key, node);
            }
        }
        return object;
    }

    public static boolean allMatch(JsonNode array, String expectedType) {
        if (array == null || !array.isArray()) {
            return false;
        }
        ArrayNode arrayNode = (ArrayNode) array;
        for (int i = 0; i < arrayNode.size(); i++) {
            JsonNode item = arrayNode.get(i);
            if ("string".equals(expectedType)) {
                if (!item.isTextual()) return false;
            } else if ("boolean".equals(expectedType)) {
                if (!item.isBoolean()) return false;
            } else if ("number".equals(expectedType)) {
                if (!item.isNumber()) return false;
            } else if ("integer".equals(expectedType)) {
                if (!item.isInt()) return false;
            } else if ("object".equals(expectedType)) {
                if (!item.isObject()) return false;
            } else if ("any".equals(expectedType)) {
                if (item.isNull()) return false;
            } else {
                return false;
            }
        }
        return true;
    }

    public static boolean allValuesMatch(ObjectNode obj, String expectedType) {
        if (obj == null) {
            return false;
        }
        for (String fieldName : keys(obj)) {
            JsonNode item = obj.get(fieldName);
            if ("string".equals(expectedType)) {
                if (!item.isTextual()) return false;
            } else if ("boolean".equals(expectedType)) {
                if (!item.isBoolean()) return false;
            } else if ("number".equals(expectedType)) {
                if (!item.isNumber()) return false;
            } else if ("integer".equals(expectedType)) {
                if (!item.isInt()) return false;
            } else if ("object".equals(expectedType)) {
                if (!item.isObject()) return false;
            } else if ("any".equals(expectedType)) {
                if (item.isNull()) return false;
            } else {
                return false;
            }
        }
        return true;
    }

    public static void addToArray(ArrayNode array, JsonNode value) {
        array.add(value);
    }

    public static boolean isString(JsonNode value) {
        if (value == null) {
            return false;
        }
        return value.isTextual();
    }

    public static boolean isJsonNode(JsonNode value) {
        if (value == null) {
            return false;
        }
        return true;
    }

    public static boolean isObjectNode(JsonNode value) {
        if (value == null) {
            return false;
        }
        return value.isObject();
    }

    public static String toString(JsonNode value) {
        return value.asText();
    }

    public static JsonNode toJsonNode(JsonNode value) {
        return value;
    }

    public static boolean isBoolean(JsonNode value) {
        if (value == null) {
            return false;
        }
        return value.isBoolean();
    }

    public static Boolean toBoolean(JsonNode value) {
        return value.asBoolean();
    }

    public static boolean isNumber(JsonNode value) {
        if (value == null) {
            return false;
        }
        return value.isNumber();
    }

    public static Number toNumber(JsonNode value) {
        if (value.isInt()) {
            return value.asInt();
        }
        if (value.isLong()) {
            return value.asLong();
        }
        return value.asDouble();
    }

    public static Integer toInteger(JsonNode value) {
        return value.asInt();
    }

    public static boolean isObject(JsonNode value) {
        if (value == null) {
            return false;
        }
        return value.isObject();
    }

    public static boolean isObjectWithProperty(JsonNode value, String propertyName) {
        if (value == null) {
            return false;
        }
        if (value.isObject()) {
            ObjectNode object = (ObjectNode) value;
            return object.has(propertyName);
        }
        return false;
    }

    public static boolean isObjectWithPropertyValue(JsonNode value, String propertyName, String propertyValue) {
        if (value == null) {
            return false;
        }
        if (value.isObject()) {
            ObjectNode object = (ObjectNode) value;
            if (object.has(propertyName)) {
                JsonNode pvalue = object.get(propertyName);
                if (!pvalue.isNull() && pvalue.isTextual()) {
                    String val = pvalue.asText();
                    return propertyValue.equals(val);
                }
            }
        }
        return false;
    }

    public static ObjectNode toObject(JsonNode value) {
        return (ObjectNode) value;
    }

    public static boolean isArray(JsonNode value) {
        if (value == null) {
            return false;
        }
        return value.isArray();
    }

    public static ArrayNode toArray(JsonNode value) {
        return (ArrayNode) value;
    }

    public static List<JsonNode> toList(JsonNode value) {
        List<JsonNode> rval = new LinkedList<>();
        ArrayNode array = (ArrayNode) value;
        for (int idx = 0; idx < array.size(); idx++) {
            JsonNode node = array.get(idx);
            rval.add(node);
        }
        return rval;
    }

    private static class PrettyPrinter extends MinimalPrettyPrinter {
        private static final long serialVersionUID = -4446121026177697380L;

        private int indentLevel = 0;

        /**
         * @see com.fasterxml.jackson.core.util.MinimalPrettyPrinter#writeStartObject(com.fasterxml.jackson.core.JsonGenerator)
         */
        @Override
        public void writeStartObject(JsonGenerator g) throws IOException {
            super.writeStartObject(g);
            indentLevel++;
            g.writeRaw("\n");
        }

        /**
         * @see com.fasterxml.jackson.core.util.MinimalPrettyPrinter#writeEndObject(com.fasterxml.jackson.core.JsonGenerator,
         *      int)
         */
        @Override
        public void writeEndObject(JsonGenerator g, int nrOfEntries) throws IOException {
            indentLevel--;
            g.writeRaw("\n");
            writeIndent(g);
            super.writeEndObject(g, nrOfEntries);
        }

        /**
         * @see com.fasterxml.jackson.core.util.MinimalPrettyPrinter#writeStartArray(com.fasterxml.jackson.core.JsonGenerator)
         */
        @Override
        public void writeStartArray(JsonGenerator g) throws IOException {
            super.writeStartArray(g);
            indentLevel++;
        }

        /**
         * @see com.fasterxml.jackson.core.util.MinimalPrettyPrinter#writeEndArray(com.fasterxml.jackson.core.JsonGenerator,
         *      int)
         */
        @Override
        public void writeEndArray(JsonGenerator g, int nrOfValues) throws IOException {
            g.writeRaw("\n");
            indentLevel--;
            writeIndent(g);
            super.writeEndArray(g, nrOfValues);
        }

        /**
         * @see com.fasterxml.jackson.core.util.MinimalPrettyPrinter#beforeObjectEntries(com.fasterxml.jackson.core.JsonGenerator)
         */
        @Override
        public void beforeObjectEntries(JsonGenerator g) throws IOException {
            writeIndent(g);
        }

        /**
         * @see com.fasterxml.jackson.core.util.MinimalPrettyPrinter#beforeArrayValues(com.fasterxml.jackson.core.JsonGenerator)
         */
        @Override
        public void beforeArrayValues(JsonGenerator g) throws IOException {
            g.writeRaw("\n");
            writeIndent(g);
        }

        /**
         * @see com.fasterxml.jackson.core.util.MinimalPrettyPrinter#writeArrayValueSeparator(com.fasterxml.jackson.core.JsonGenerator)
         */
        @Override
        public void writeArrayValueSeparator(JsonGenerator g) throws IOException {
            super.writeArrayValueSeparator(g);
            g.writeRaw("\n");
            writeIndent(g);
        }

        /**
         * @see com.fasterxml.jackson.core.util.MinimalPrettyPrinter#writeObjectEntrySeparator(com.fasterxml.jackson.core.JsonGenerator)
         */
        @Override
        public void writeObjectEntrySeparator(JsonGenerator g) throws IOException {
            super.writeObjectEntrySeparator(g);
            g.writeRaw("\n");
            writeIndent(g);
        }

        /**
         * @see com.fasterxml.jackson.core.util.MinimalPrettyPrinter#writeObjectFieldValueSeparator(com.fasterxml.jackson.core.JsonGenerator)
         */
        @Override
        public void writeObjectFieldValueSeparator(JsonGenerator g) throws IOException {
            super.writeObjectFieldValueSeparator(g);
            g.writeRaw(" ");
        }

        private void writeIndent(JsonGenerator g) throws IOException {
            for (int idx = 0; idx < this.indentLevel; idx++) {
                g.writeRaw("    ");
            }
        }
    }

}

File: data-models/src/main/ts/src/io/apitomy/datamodels/models/util/JsonUtil.ts

export class JsonUtil {

    public static keys(json: object): string[] {
        if (!json) {
            return [];
        }
        return Object.keys(json);
    }

    public static matchingKeys(regex: string, json: object): string[] {
        const re: RegExp = new RegExp(regex);
        return JsonUtil.keys(json).filter(key => re.test(key));
    }

    public static getProperty(json: object, propertyName: string): any {
        let rval: any = json[propertyName];
        if (rval === undefined) {
            rval = null;
        }
        return rval;
    }

    public static setProperty(json: object, propertyName: string, propertyValue: any): void {
        if (propertyValue !== null) {
            json[propertyName] = propertyValue;
        }
    }

    public static consumeProperty(json: any, propertyName: string): any {
        let rval: any = JsonUtil.getProperty(json, propertyName);
        if (rval) {
            delete json[propertyName];
        }
        return rval;
    }

    public static stringify(json: any): string {
        return JSON.stringify(json);
    }

    public static removeProperty(json: any, propertyName: string): void {
        if (json != null && typeof json === "object") {
            delete json[propertyName];
        }
    }

    public static parseJSON(jsonString: string): any {
        return JSON.parse(jsonString);
    }

    public static clone(json: any): any {
        return JSON.parse(JSON.stringify(json));
    }

    public static collectionToList<T>(collection: T[]): T[] {
        if (!collection) {
            return [];
        }
        return [...collection];
    }

    public static objectNode(): any {
        return {};
    }

    public static textNode(value: string): any {
        return value;
    }

    public static arrayNode(): any {
        return [];
    }

    public static toJsonNode(value: any): any {
        return value;
    }

    // JSweet mangled names for toJsonNode overloads
    public static toJsonNode$java_lang_Object(value: any): any {
        return value;
    }

    public static toJsonNode$com_fasterxml_jackson_databind_JsonNode(value: any): any {
        return value;
    }

    public static toArrayNode(list: any[]): any[] {
        if (list == null) {
            return null;
        }
        const array: any[] = [];
        for (let i: number = 0; i < list.length; i++) {
            const node: any = list[i];
            if (node != null) {
                array.push(node);
            }
        }
        return array;
    }

    public static toObjectNode(value: any): object {
        return value;
    }

    public static allMatch(array: any, expectedType: JsonUtil.JsonType): boolean {
        if (array == null || !Array.isArray(array)) {
            return false;
        }
        for (let i: number = 0; i < array.length; i++) {
            if (!expectedType.matches(array[i])) return false;
        }
        return true;
    }

    public static allValuesMatch(obj: any, expectedType: JsonUtil.JsonType): boolean {
        if (obj == null) {
            return false;
        }
        const fieldNames: string[] = JsonUtil.keys(obj);
        for (let i: number = 0; i < fieldNames.length; i++) {
            if (!expectedType.matches(obj[fieldNames[i]])) return false;
        }
        return true;
    }

    public static addToArray(array: Array<any>, value: any): void {
        array.push(value);
    }

    public static isString(value: any): boolean {
        if (value == null) {
            return false;
        }
        return typeof value === "string";
    }

    public static isJsonNode(value: any): boolean {
        if (value == null) {
            return false;
        }
        return true;
    }

    public static isObjectNode(value: any): boolean {
        if (value == null) {
            return false;
        }
        return typeof value === "object";
    }

    public static toString(value: any): string {
        return value;
    }

    public static isBoolean(value: any): boolean {
        if (value == null) {
            return false;
        }
        return typeof value === "boolean";
    }

    public static toBoolean(value: any): boolean {
        return value;
    }

    public static isNumber(value: any): boolean {
        if (value == null) {
            return false;
        }
        return typeof value === "number";
    }

    public static toNumber(value: any): number {
        return value;
    }

    public static toInteger(value: any): number {
        return value;
    }

    public static isObject(value: any): boolean {
        if (value == null) {
            return false;
        }
        return typeof value === "object" && !Array.isArray(value);
    }

    public static isObjectWithProperty(value: any, propertyName: string): boolean {
        if (JsonUtil.isObject(value)) {
            return value.hasOwnProperty(propertyName);
        }
        return false;
    }

    public static isObjectWithStringPropertyValue(value: any, propertyName: string, propertyValue: string): boolean {
        if (JsonUtil.isObject(value)) {
            if (value.hasOwnProperty(propertyName)) {
                const pvalue: any = value[propertyName];
                if (pvalue != null && typeof pvalue === "string") {
                    return propertyValue === pvalue;
                }
            }
        }
        return false;
    }

    public static toObject(value: any): object {
        return value;
    }

    public static isArray(value: any): boolean {
        if (value == null) {
            return false;
        }
        return Array.isArray(value);
    }

    public static toArray(value: any): any[] {
        return value;
    }

    public static toList(value: any): any[] {
        return value;
    }

}

export namespace JsonUtil {
    export class JsonType {
        public static STRING: JsonType = new JsonType((v: any) => typeof v === "string");
        public static BOOLEAN: JsonType = new JsonType((v: any) => typeof v === "boolean");
        public static NUMBER: JsonType = new JsonType((v: any) => typeof v === "number");
        public static INTEGER: JsonType = new JsonType((v: any) => typeof v === "number" && Number.isInteger(v));
        public static OBJECT: JsonType = new JsonType((v: any) => typeof v === "object" && !Array.isArray(v));
        public static ANY: JsonType = new JsonType((v: any) => v != null);

        private readonly matcher: (value: any) => boolean;

        private constructor(matcher: (value: any) => boolean) {
            this.matcher = matcher;
        }

        public matches(node: any): boolean {
            return this.matcher(node);
        }
    }
}
